### ISSUE TITLE
Storing object location

### ISSUE THREAD

**Issue by Kludex** [Instagram/LibCST#706]
Hi there 👋 

I have a question more related to best practice/how-to, I hope you guys can help me. 🙏 

I'm working on an opinionated tool, that refactors the function parameters and the call for that function through the source code. As an example, if I have a function with a lot of parameters, I want that function to have only keyword only parameters, like:
```python
def do_something(a, b, c, d):
    ...

do_something(1, True, "potato", None)
```
Turns into:
```python
def do_something(*, a, b, c, d):
    ...

do_something(a=1, b=True, c="potato", d=None)
```
I'm using a transformer to modify the function definitions first (while I store the parameters), and then I use another transformer for the calls, using the stored parameters as argument to the latter transformer. See more about it [here](https://github.com/Kludex/kwonly-transformer/blob/main/kwonly_transformer/main.py).

What I'm not sure about is how to store the location of that function. If I have two functions with the same name in the source code, I don't like that to have a naming conflict. If I use a metadata to get the location, I still have the problem of internal functions, as those will be stored as `some_module.<locals>.function_name` (I mean, maybe this is not a problem and the name reference here is just fine).

Anyway, I'd like to know if there's any reference on how to achieve this, and if my thoughts make sense. Can someone help me?

Thanks for the package, btw! 🙏

**Comment by zsol** [comment]
I would use `FullyQualifiedNameProvider` for this. As you mentioned, I also don't see why `some_module.<locals>.function_name` is a problem in this case - all you need is a handle that represents the function at runtime so you can connect the invocation with the definition.

You could also use `ScopeProvider` and for each `Call` node, get the `Access` for the `call.func` node and connect that back to the original function def, but that will only ever work in the same file. I prefer the qualified name approach because it's both simpler, and will work across files when you eventually get there.

**Comment by Kludex** [comment]
I've tried to use the `FullyQualifiedNameProvider` some days ago. I think I was having some "issue" in which I was not able to match the function called to the actual function definition "path" (e.g. `some_module.<locals>.function_name`).

Do you perhaps have an analogous implementation somewhere I can check? In any case, I'll try again over the weekend. Thanks! 🙏

**Comment by Kludex** [comment]
Assume I have two files:
```python
# potato/main.py
from potato.another import Potato, home

home(1, 2, 3)
var = Potato()
var.home(1, 2, 3)

# potato/another.py
def home(a, b, c):
    ...


class Potato:
    def home(self, a, b, c):
        ...
```
The `QualifiedName` repr of the calls will be:
```python
{QualifiedName(name='potato.another.home', source=<QualifiedNameSource.IMPORT: 1>)}
{QualifiedName(name='potato.another.Potato', source=<QualifiedNameSource.IMPORT: 1>)}
{QualifiedName(name='potato.main.var.home', source=<QualifiedNameSource.LOCAL: 3>)}
```
The `QualifiedName` repr of the `FunctionDef` will be:
```python
{QualifiedName(name='potato.another.home', source=<QualifiedNameSource.LOCAL: 3>)}
{QualifiedName(name='potato.another.Potato.home', source=<QualifiedNameSource.LOCAL: 3>)}
```

For `potato.another.home` (`Call`) I can link to the same name (`FunctionDef`), but for `potato.main.var.home`, I can't link to `potato.another.Potato.home`. The idea here is that I would need to replace the `main.var` by `another.Potato` or is there an alternative?

**Comment by zsol** [comment]
Yeah this scenario is too advanced for a simple tool like `ScopeProvider`. It requires a bit more static analysis.

**Comment by Kludex** [comment]
Do you know any code reference of someone doing something like this?

If not... Well... I guess this issue can be closed. Thanks for the help! 🙏

**Comment by zsol** [comment]
Well, type checkers can do this. You might have some luck with TypeInferenceProvider (which uses pyre), but honestly it's painful to set up.

Good luck!
